File: nikkake_flutter/lib/domain/date_utils.dart
import '../models/models.dart';

/// 日付と表示フォーマット。
/// すべて純粋関数なのでストレージ無しで単体テストできる。

String getDateString([DateTime? date]) {
  final d = date ?? DateTime.now();
  return '${d.year.toString().padLeft(4, '0')}-'
      '${d.month.toString().padLeft(2, '0')}-'
      '${d.day.toString().padLeft(2, '0')}';
}

/// 'YYYY-MM-DD' をローカルタイムの0時として解釈する
DateTime parseDateString(String value) {
  final parts = value.split('-').map(int.parse).toList();
  return DateTime(parts[0], parts.length > 1 ? parts[1] : 1, parts.length > 2 ? parts[2] : 1);
}

DateTime startOfDay(DateTime date) => DateTime(date.year, date.month, date.day);

/// 時刻を無視した日数差。夏時間で±1時間ずれても丸めで吸収する
int daysBetween(DateTime from, DateTime to) =>
    (startOfDay(to).difference(startOfDay(from)).inHours / 24).round();

DateTime addDays(DateTime date, int days) =>
    DateTime(date.year, date.month, date.day + days, date.hour, date.minute, date.second);

/// そのルーティンを今日やる予定かどうか。
/// 「今日もう終わったか」は別で判定するので、ここでは予定だけを見る。
bool isRoutineDueToday(Routine routine, RoutineLog? lastLog, [DateTime? now]) {
  if (!routine.isActive) return false;
  final today = now ?? DateTime.now();

  switch (routine.frequencyType) {
    case FrequencyType.daily:
      return true;

    case FrequencyType.weekly:
      // 曜日が未設定なら毎日扱いにする
      if (routine.frequencyDays.isEmpty) return true;
      // Dartの weekday は 月=1..日=7。DBは 日=0..土=6 なので合わせる
      final dbWeekday = today.weekday % 7;
      return routine.frequencyDays.contains(dbWeekday);

    case FrequencyType.everyNDays:
      if (lastLog == null) return true; // 一度もやっていなければ今日から始める
      final interval = routine.frequencyValue < 1 ? 1 : routine.frequencyValue;
      return daysBetween(parseDateString(lastLog.logDate), today) >= interval;
  }
}

String formatDuration(int? seconds) {
  if (seconds == null) return '--:--';
  final total = seconds < 0 ? 0 : seconds;
  final h = total ~/ 3600;
  final m = (total % 3600) ~/ 60;
  final s = total % 60;

  if (h > 0) {
    return '$h:${m.toString().padLeft(2, '0')}:${s.toString().padLeft(2, '0')}';
  }
  return '$m:${s.toString().padLeft(2, '0')}';
}

String formatWeight(double? weight) {
  if (weight == null) return '-- kg';
  // 小数第1位まで。.0 は省く
  final rounded = double.parse(weight.toStringAsFixed(1));
  final text = rounded == rounded.roundToDouble() ? rounded.toInt().toString() : rounded.toString();
  return '$text kg';
}

const _dayNames = ['日', '月', '火', '水', '木', '金', '土'];

String formatFrequency(Routine routine) {
  switch (routine.frequencyType) {
    case FrequencyType.daily:
      return '毎日';
    case FrequencyType.everyNDays:
      return routine.frequencyValue <= 1 ? '毎日' : '${routine.frequencyValue}日ごと';
    case FrequencyType.weekly:
      if (routine.frequencyDays.isEmpty) return '毎日';
      final sorted = [...routine.frequencyDays]..sort();
      return sorted.map((d) => _dayNames[d]).join('・');
  }
}

String greetingForHour(int hour) {
  if (hour < 5) return 'おつかれさま';
  if (hour < 12) return 'おはよう！';
  if (hour < 18) return 'こんにちは！';
  return 'こんばんは！';
}

File: nikkake_kmp/shared/src/commonMain/kotlin/com/myapplication/common/domain/DateUtils.kt
package com.myapplication.common.domain

import com.myapplication.common.data.FrequencyType
import com.myapplication.common.data.Routine
import com.myapplication.common.data.RoutineLog
import kotlinx.datetime.Clock
import kotlinx.datetime.DatePeriod
import kotlinx.datetime.LocalDate
import kotlinx.datetime.TimeZone
import kotlinx.datetime.daysUntil
import kotlinx.datetime.plus
import kotlinx.datetime.toLocalDateTime

/**
 * 日付と表示フォーマット。
 * すべて純粋関数なのでストレージ無しで単体テストできる。
 */

fun today(timeZone: TimeZone = TimeZone.currentSystemDefault()): LocalDate =
    Clock.System.now().toLocalDateTime(timeZone).date

fun nowIso(): String = Clock.System.now().toString()

fun getDateString(date: LocalDate = today()): String {
    val month = date.monthNumber.toString().padStart(2, '0')
    val day = date.dayOfMonth.toString().padStart(2, '0')
    return "${date.year}-$month-$day"
}

fun parseDateString(value: String): LocalDate = LocalDate.parse(value)

fun daysBetween(from: LocalDate, to: LocalDate): Int = from.daysUntil(to)

fun addDays(date: LocalDate, days: Int): LocalDate = date.plus(DatePeriod(days = days))

/**
 * DBは 日=0..土=6 で曜日を持つ。
 * DayOfWeek.ordinal は 月=0..日=6 なので +1 して 7 で割った余りに落とす。
 */
fun dbWeekday(date: LocalDate): Int = (date.dayOfWeek.ordinal + 1) % 7

/**
 * そのルーティンを今日やる予定かどうか。
 * 「今日もう終わったか」は別で判定するので、ここでは予定だけを見る。
 */
fun isRoutineDueToday(routine: Routine, lastLog: RoutineLog?, today: LocalDate = today()): Boolean {
    if (!routine.isActive) return false

    return when (routine.frequencyType) {
        FrequencyType.DAILY -> true

        // 曜日が未設定なら毎日扱いにする
        FrequencyType.WEEKLY ->
            routine.frequencyDays.isEmpty() || routine.frequencyDays.contains(dbWeekday(today))

        FrequencyType.EVERY_N_DAYS -> {
            // 一度もやっていなければ今日から始める
            if (lastLog == null) true
            else daysBetween(parseDateString(lastLog.logDate), today) >= maxOf(1, routine.frequencyValue)
        }

        else -> false
    }
}

fun formatDuration(seconds: Int?): String {
    if (seconds == null) return "--:--"

    val total = maxOf(0, seconds)
    val h = total / 3600
    val m = (total % 3600) / 60
    val s = total % 60

    return if (h > 0) {
        "$h:${m.toString().padStart(2, '0')}:${s.toString().padStart(2, '0')}"
    } else {
        "$m:${s.toString().padStart(2, '0')}"
    }
}

fun formatWeight(weight: Double?): String {
    if (weight == null) return "-- kg"
    return "${formatWeightNumber(weight)} kg"
}

/** 単位なしの重量。小数第1位まで、.0 は省く */
fun formatWeightNumber(weight: Double): String {
    val rounded = kotlin.math.round(weight * 10) / 10
    return if (rounded == kotlin.math.floor(rounded)) rounded.toInt().toString() else rounded.toString()
}

/** 前回表示に必要な最小限。ExerciseLog からも API 応答からも作れる */
data class PreviousSetLike(
    val reps: Int? = null,
    val weight: Double? = null,
    val durationSec: Int? = null,
)

/**
 * ワークアウト画面の「前回: …」に出す文字列。
 *
 * packages/contract/domain_cases.json の formatPreviousSets が正で、
 * Ruby の Domain::Stats.format_previous_sets と同じ答えを返さなければならない。
 */
fun formatPreviousSets(sets: List<PreviousSetLike>): String = sets.joinToString(" / ") { set ->
    if (set.durationSec != null) {
        "${set.durationSec}秒"
    } else {
        val label = set.weight?.let { formatWeightNumber(it) } ?: "自重"
        "$label×${set.reps ?: "-"}"
    }
}

private val dayNames = listOf("日", "月", "火", "水", "木", "金", "土")

fun formatFrequency(routine: Routine): String = when (routine.frequencyType) {
    FrequencyType.DAILY -> "毎日"
    FrequencyType.EVERY_N_DAYS ->
        if (routine.frequencyValue <= 1) "毎日" else "${routine.frequencyValue}日ごと"
    FrequencyType.WEEKLY ->
        if (routine.frequencyDays.isEmpty()) "毎日"
        else routine.frequencyDays.sorted().joinToString("・") { dayNames[it] }
    else -> ""
}

fun greetingForHour(hour: Int): String = when {
    hour < 5 -> "おつかれさま"
    hour < 12 -> "おはよう！"
    hour < 18 -> "こんにちは！"
    else -> "こんばんは！"
}

fun currentHour(timeZone: TimeZone = TimeZone.currentSystemDefault()): Int =
    Clock.System.now().toLocalDateTime(timeZone).hour
